### PR TITLE
[State Observability] Ray log alpha API

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

This is the PR to implement `ray log` to the server side. The PR is continued from https://github.com/ray-project/ray/pull/24068. 

The PR supports two endpoints;

```
/api/v0/logs # list logs of the node id filtered by the given glob. 
/api/v0/logs/{[file | stream]}?filename&pid&actor_id&task_id&interval&lines # Stream the requested file log. The filename can be inferred by pid/actor_id/task_id
```

Some tests need to be re-written, I will do it soon.

As a follow-up after this PR, there will be 2 PRs.
- PR to add actual CLI
- PR to remove in-memory cached logs and do on-demand query for actor/worker logs

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rkooo567** [comment]
This PR is ready to review.

File: dashboard/modules/log/log_manager.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+                    raise ValueError(
+                        f"Worker ID for Actor ID {options.actor_id} not found. "
+                        "Actor is not scheduled yet."
+                    )
+
+                log_files = await self.list_logs(
+                    node_id, options.timeout, glob_filter=f"*{worker_id}*"
+                )
+                for file in log_files["worker_out"]:
+                    if file.split(".")[0].split("-")[1] == worker_id:
```
Comment (scv119) [inline dashboard/modules/log/log_manager.py:126]: nit some inline comments would be helpful.

File: dashboard/modules/log/log_manager.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+from ray.experimental.state.state_manager import StateDataSourceClient
+
+# TODO(sang): Remove the usage of this class.
+from ray.dashboard.datacenter import DataSource
+
+
+logger = logging.getLogger(__name__)
+
+
+class LogsManager:
```
Comment (scv119) [inline dashboard/modules/log/log_manager.py:19]: Do we have unit test for LogsManager?

File: dashboard/modules/state/state_head.py [hunk 3]
```diff
@@ -126,33 +138,115 @@ async def list_jobs(self, req) -> aiohttp.web.Response:
...
     @routes.get("/api/v0/runtime_envs")
-    async def list_runtime_envs(self, req) -> aiohttp.web.Response:
+    async def list_runtime_envs(self, req: aiohttp.web.Request) -> aiohttp.web.Response:
         return await self._handle_list_api(self._state_api.list_runtime_envs, req)
 
+    @routes.get("/api/v0/logs")
+    async def list_logs(self, req: aiohttp.web.Request) -> aiohttp.web.Response:
+        """Return a list of log files on a given node id.
+
+        Note that the semantic of this "list" is different from other
```
Comment (scv119) [inline dashboard/modules/state/state_head.py:170]: huh, clarify a bit more?

**Review by scv119 — APPROVED** [review]
mostly just nits.

**Review by rkooo567 — COMMENTED** [review]
Rewrote the ugly resolve_file_and_node func

**Pull request merged**